### PR TITLE
Add OrchardCore.Sms module and add 2FA method via SMS feature

### LINKED ISSUE #14127: Add SMS support
We need a way to be able to use SMS service like Twilio or other providers.

### LINKED ISSUE #14128: Add two-factor authentication using SMS service
Add a way to send two-factor authentication token via SMS.

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Fix #14127 
Fix #14128

# New module `OrchardCore.Sms` was added.

image

image

### Console Provider
This provider is only available in Development environment for debugging purposes.

image

### Twilio Provider
image

#### Custom Provider?
Sure, you can add any custom provider you like. You can simply do that by implementing the `ISmsProvider`
```
services.AddSmsProvider<YourCustomImplemenation>("Name for your implementation")
```
## Two-Factor Authentication
New feature called `Two-Factor SMS Method` was added to allow the admin to utilize the `OrchardCore.Sms` feature to send one-time 2FA code to the phone number using SMS message.

image

image

image

## Workflow support
image

## Notifications
New feature called "SMS Notifications" was added to allow the use to send notifications via SMS provider.
image

### CODE REVIEW

**Comment by MikeAlhayek** [comment]
@idan-h this PR may be an interest to you as I think you were trying to accomplish 2FA using SMS.

**Comment by idan-h** [comment]
Looks great

How would we go about adding a custom sms provider?

We could make a setting that lets the user input request details (get/post, url, content etc..)

And then utilize this to make the sms request for any local or custom provider.

Let me know what you think, and if you understood what I mean or you need clarification :)

**Comment by MikeAlhayek** [comment]
@idan-h I don't quite understand what you want to achieve. You can easily add a new provider. Please refer to the documentation file included in this PR to see how you can easily add your own SMS provider.

**Comment by idan-h** [comment]
> @idan-h I don't quite understand what you want to achieve. You can easily add a new provider. Please refer to the documentation file included in this PR to see how you can easily add your own SMS provider.

What I mean is a generic http provider, that lets you setup a custom sms provider without any code change (setup the request details, including request type, url, content, headers etc)

**Comment by MikeAlhayek** [comment]
It sounds like you are defining an implementation for a custom provider. You can so that like this

```
public class HttpSmsImplemenation : ISmsProvider
{
   public Task<SmsResult> SendAsync(SmsMessage)
   {
    // add your custom logic
    }
}
```

then register it
```
services.AddSmsProvider<HttpSmsImplemenation >("Http Request provider")
```

File: src/OrchardCore.Modules/OrchardCore.Sms/Views/TwilioSettings.Edit.cshtml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+@model TwilioSettingsViewModel
+
+<div class="sms-settings-container d-none" id="SmsSettingsContainer@(SmsConstants.TwilioServiceName)">
+    <h4>@T["Twilio Account Info"]</h4>
+    <div class="mb-3" asp-validation-class-for="PhoneNumber">
+        <label asp-for="PhoneNumber">@T["Phone Number"]</label>
+        <input asp-for="PhoneNumber" class="form-control" />
+        <span asp-validation-for="PhoneNumber"></span>
+        <span>@T["Phone number must include a country code. For example, +1 for United States."]</span>
```
Comment (agriffard) [inline src/OrchardCore.Modules/OrchardCore.Sms/Views/TwilioSettings.Edit.cshtml:12]: Please use `class="hint"` in help messages.

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Sms/Views/TwilioSettings.Edit.cshtml:12]: @agriffard done!

File: src/docs/reference/modules/Sms/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# SMS (`OrchardCore.Sms`)
+
+This module provides the infrastructure necessary to send messages using `SMS` service.
+
+## SMS Settings
+
+Enabling the `OrchardCore.Sms` will add a new settings under `Configurations` >> `Settings` >> `SMS`. You can utilize these settings to set up the default SMS provider configuration. The following are the providers that are readily accessible.
```
Comment (agriffard) [inline src/docs/reference/modules/Sms/README.md:7]: Enabling the `OrchardCore.Sms` will add a new settings under
=>
Enabling the `OrchardCore.Sms` feature will add a new settings page under

File: src/docs/reference/modules/Sms/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+| Provider | Description |
+| --- | --- |
+| `Console` | This particular provider is exclusively meant for debugging purposes and should never be employed in a production environment. It permits the message to be displayed on the debug console. |
```
Comment (agriffard) [inline src/docs/reference/modules/Sms/README.md:11]: employed = used or you meant `deployed` ?

File: src/docs/reference/modules/Sms/README.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+| `Twilio` | Opting for this provider enables the utilization of Twilio service for sending SMS messages. By choosing this provider, you will need to input your Twilio account settings. |
```
Comment (agriffard) [inline src/docs/reference/modules/Sms/README.md:12]: to input your = `to enter your` ?

File: src/OrchardCore.Build/Dependencies.props [hunk 5]
```diff
@@ -26,6 +26,7 @@
     <PackageManagement Include="GraphQL.SystemTextJson" Version="4.8.0" />
     <PackageManagement Include="Jint" Version="3.0.0-beta-2049" />
     <PackageManagement Include="HtmlSanitizer" Version="8.0.692" />
+    <PackageManagement Include="libphonenumber-csharp" Version="8.13.18" />
```
Comment (agriffard) [inline src/OrchardCore.Build/Dependencies.props:29]: Please add new dependencies versions and licenses to src\docs\resources\libraries\Readme.md

File: src/docs/reference/modules/Sms/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/docs/reference/modules/Sms/README.md:7]: yes. meaning it'll be ready and accessible.

File: src/docs/reference/modules/Sms/README.md (hunk 4 above)
Comment (MikeAlhayek) [reply src/docs/reference/modules/Sms/README.md:12]: same thing

File: src/docs/reference/modules/Sms/README.md (hunk 2 above)
Comment (MikeAlhayek) [reply src/docs/reference/modules/Sms/README.md]: Done.

File: src/OrchardCore.Build/Dependencies.props (hunk 5 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Build/Dependencies.props:29]: added.

File: src/docs/reference/modules/Sms/README.md (hunk 3 above)
Comment (MikeAlhayek) [reply src/docs/reference/modules/Sms/README.md]: changed

File: src/OrchardCore/OrchardCore.Sms.Core/DefaultPhoneFormatValidator.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        {
+            var phone = _phoneNumberUtil.Parse(phoneNumber, null);
+
+            return _phoneNumberUtil.IsValidNumber(phone);
+        }
+        catch
+        {
+            return false;
+        }
+    }
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Sms.Core/DefaultPhoneFormatValidator.cs:23]:
```suggestion
    public bool IsValid(string phoneNumber)
    {
        var phoneNumberUtil = PhoneNumberUtil.GetInstance();

        try
        {
            var phone = phoneNumberUtil.Parse(phoneNumber, null);

            return phoneNumberUtil.IsValidNumber(phone);
        }
        catch
        {
            return false;
        }
    }
```

**Review by sebastienros — APPROVED** [review]

File: src/OrchardCore/OrchardCore.Sms.Core/Services/SmsProviderFactory.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
+using System;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using OrchardCore.Entities;
+using OrchardCore.Settings;
+
+namespace OrchardCore.Sms.Services;
+
+public class SmsProviderFactory : ISmsProviderFactory
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Sms.Core/Services/SmsProviderFactory.cs:10]: Simplify

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.Sms.Core/Services/SmsProviderFactory.cs]: @sebastienros I removed the factory and simplified it. Let me know your thoughts

https://github.com/OrchardCMS/OrchardCore/pull/14125/files#diff-cc9687e2af5bb116baccc4fab6fe03cb0d9947eef620be03fc44cc5e865f71faR15-R33

**Review by hishamco — CHANGES_REQUESTED** [review]
I prefer to use `OC.Sms.Twilio` rather than `OC.Sms.Core`, coz there will be many implementation as SMS providers

**Comment by MikeAlhayek** [comment]
@hishamco that would be a good name if we add a module for each implementation. The way the SMS module is designed is that you can configure any service you want not just Twilio. Adding a new implementation is as simple as registering a new type. For example, if you want to have a new implementation for "MessageBird", you don't need a new feature or module for that. All you do is provide the implementation and register it. The implementation will then become available in the configuration option for you to select from. So here you don't need to enable new feature or new module to bring in your own implementation.

**Comment by hishamco** [comment]
> if you want to have a new implementation for "MessageBird", you don't need a new feature or module for that. All you do is provide the implementation and register it. The implementation will then become available in the configuration option for you to select from.

This new implementation needs to be added to a module, right?

**Comment by MikeAlhayek** [comment]
Yes. It can be added to any more like directly into the Sms module. New implementation don't need their own module or feature.

**Comment by hishamco** [comment]
Still I didn't understand, I'm thinking for a dev that are using OC NuGet packages not the source code

**Comment by MikeAlhayek** [comment]
yes sure. you can create a custom module/feature in your own project. add dependency to `OrchardCore.Sms` then create an implementation of another provider like this for example


public class MessageBirdProvider : ISmsProvider
{
   public Task<SmsResult> SendAsync(SmsMessage)
   {
        // add your custom logic
    }
}

then register it

```
services.AddSmsProvider<MessageBirdProvider>("MessageBird");
```

Now, if you go to `Configurations` >> `Settings` >> `SMS`, you'll see an option to  select **MessageBird** as an options.

You may want to create a custom settings for that provider, to do that follow that I did in `TwilioSettingsDisplayDriver` also the view should be wrapped with a div like this

```
<div class="sms-settings-container d-none" id="SmsSettingsContainerMessageBird">

</div>
```

notice the wrapper has a special class `sms-settings-container` and special id that is **SmsSettingsContainer** post fixed with your provider name. 

This way your custom settings will bu only visible when your provider is selected.

**Comment by hishamco** [comment]
As expected that's why I prefer to use `OC.Sms.Twillio` instead, we already used `OC.XXX.Core` for the default implementation that relies on ASP.NET Core for instance `OC.Users.Core` & `OC.Roles.Core`

**Comment by MikeAlhayek** [comment]
having a separate project for each provider we support sounds instance to me.

**Comment by hishamco** [comment]
It's more like Social Auth, Search Providers .. etc, because we now have a Twillio specific nothing else, right?

**Comment by MikeAlhayek** [comment]
It's not. With what we have today, if you ever need SMS services, simply add dependency on SMS feature "without worrying about implementations" using settings you can select which implementation you want to use. This is a good approach because anyone wants SMS they can easily depend on the SMS module.

**Comment by hishamco** [comment]
I need to look deeply at what you did, thanks anyway for the effort that you made

**Comment by MikeAlhayek** [comment]
Yes. Create a feature that depends on OrchardCore.Sms and implement and register your own provider


PS: why the Chinese writing?

**Comment by onestar1** [comment]
> Yes. Create a feature that depends on OrchardCore.Sms and implement and register your own provider

> 

> PS: why the Chinese writing?

Thanks, seems to be the browser reason.

**Pull request merged**